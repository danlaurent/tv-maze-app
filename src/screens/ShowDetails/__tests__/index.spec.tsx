import React from 'react';
import {
  render,
  RenderAPI,
  fireEvent,
  waitFor,
} from '@testing-library/react-native';
import ShowDetailsScreen from '..';

const navigationMock: any = {
  navigate: jest.fn(),
  goBack: jest.fn(),
};

const routeMock: any = {
  params: {
    showId: 1,
  },
};

const mockShowResponse = {
  id: 1,
  name: 'testSuccess',
  genres: ['Drama', 'Science-Fiction', 'Thriller'],
  _embedded: {
    episodes: [
      {
        id: 1,
        season: 1,
        name: 'testSeason',
      },
    ],
  },
};

jest.mock('../../../hooks/useFetch', () => ({
  useFetch: () => ({
    data: mockShowResponse,
    loading: true,
  }),
}));

describe('ShowDetailsScreen', () => {
  let tree: RenderAPI;
  beforeEach(() => {
    tree = render(
      <ShowDetailsScreen navigation={navigationMock} route={routeMock} />
    );
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const snapshot = tree.toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('renders the season', () => {
    expect(tree.getByTestId('ShowDetails_season1')).toBeDefined();
  });

  it('renders the loading', () => {
    expect(tree.getByTestId('ShowDetailsLoading')).toBeDefined();
  });

  describe('When the back button is pressed', () => {
    beforeEach(async () => {
      await fireEvent.press(tree.getByTestId('ShowDetailsHeader_backButton'));
    });
    it('fires goBack action', () => {
      expect(navigationMock.goBack).toBeCalledTimes(1);
    });
  });

  describe('When an episode is pressed', () => {
    beforeEach(async () => {
      await fireEvent.press(
        tree.getByTestId('ShowDetails_season1_episodeButton_1')
      );
    });
    it('navigates to the EpisodeDetails screen', () => {
      expect(navigationMock.navigate).toHaveBeenCalledWith(
        'EpisodeDetailsScreen',
        { episode: { id: 1, name: 'testSeason', season: 1 } }
      );
    });
  });
});
