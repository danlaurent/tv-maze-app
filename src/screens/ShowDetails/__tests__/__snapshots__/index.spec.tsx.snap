// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetailsScreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#191a2f",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
        }
      }
      testID="ShowDetailsHeader"
    >
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "height": null,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingHorizontal": 20,
            "paddingVertical": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="ShowDetailsHeader_backButton"
        >
          <Text />
        </View>
      </View>
    </View>
    <RCTScrollView
      ListHeaderComponent={
        <View
          style={
            Object {
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "marginBottom": 16,
                "paddingHorizontal": 20,
              }
            }
          >
            <Poster
              size="large"
            />
            <DetailsMainInfo
              subtitle="Drama | Science-Fiction | Thriller"
              title="testSuccess"
            />
          </View>
          <View
            style={
              Object {
                "marginBottom": 20,
                "paddingHorizontal": 20,
              }
            }
          >
            <Summary />
          </View>
        </View>
      }
      data={
        Array [
          Object {
            "episodes": Array [
              Object {
                "id": 1,
                "name": "testSeason",
                "season": 1,
              },
            ],
            "season": 1,
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "paddingTop": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "marginBottom": 16,
                  "paddingHorizontal": 20,
                }
              }
            >
              <View
                style={
                  Object {
                    "borderRadius": 4,
                    "height": 295,
                    "width": 210,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "stretch",
                      "backgroundColor": "#131425",
                      "borderRadius": 4,
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                  testID="undefined_imagePlaceholder"
                >
                  <Text />
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 24,
                      "fontWeight": "700",
                      "marginBottom": 8,
                      "marginTop": 16,
                    }
                  }
                >
                  testSuccess
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#6d6f7e",
                      "marginBottom": 8,
                      "textAlign": "center",
                    }
                  }
                  testID="undefined_subtitle"
                >
                  Drama | Science-Fiction | Thriller
                </Text>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text />
                  <Text
                    style={
                      Object {
                        "color": "#c6c7d2",
                        "fontSize": 16,
                        "fontWeight": "700",
                        "marginLeft": 8,
                      }
                    }
                    testID="undefined_rating"
                  >
                    -
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "marginBottom": 20,
                  "paddingHorizontal": 20,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#6d6f7e",
                    "textAlign": "center",
                  }
                }
              >
                
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            testID="ShowDetails_season1"
          >
            <View
              style={
                Object {
                  "backgroundColor": "#121322",
                  "borderBottomWidth": 1,
                  "borderColor": "#0c0d18",
                  "borderTopWidth": 1,
                  "height": 56,
                  "justifyContent": "center",
                  "paddingHorizontal": 12,
                  "paddingVertical": 4,
                }
              }
              testID="ShowDetails_season1_header"
            >
              <Text
                style={
                  Object {
                    "color": "#c6c7d2",
                    "fontSize": 16,
                    "fontWeight": "700",
                  }
                }
                testID="ShowDetails_season1_headerText"
              >
                Season 
                1
              </Text>
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#0c0d18",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderColor": "#121322",
                    "flexDirection": "row",
                    "height": 48,
                    "justifyContent": "space-between",
                    "opacity": 1,
                    "paddingHorizontal": 20,
                    "paddingVertical": 4,
                  }
                }
                testID="ShowDetails_season1_episodeButton_1"
              >
                <Text
                  numberOfLines={1}
                  style={
                    Object {
                      "color": "#6d6f7e",
                      "fontSize": 16,
                      "fontWeight": "700",
                    }
                  }
                  testID="ShowDetails_season1_episodeText_1"
                >
                   - 
                  testSeason
                </Text>
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginBottom": 32,
        }
      }
      testID="ShowDetailsLoading"
    >
      <ActivityIndicator
        animating={true}
        color="white"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
